src/java/io/bazel/rulesscala/scalac/reporter/DepsTrackingReporter.java

```diff
@@ -0,0 +1,305 @@
+import scala.tools.nsc.Settings;
+import scala.tools.nsc.reporters.ConsoleReporter;
+import scala.tools.nsc.reporters.Reporter;
+
+public class DepsTrackingReporter extends ConsoleReporter {
+
+  private static final String HJAR_JAR_SUFFIX = "-hjar.jar";
+  private static final String IJAR_JAR_SUFFIX = "-ijar.jar";
+  private final Set<String> usedJars = new HashSet<>();
+
+  private final Map<String, String> jarToTarget = new HashMap<>();
+  private final Map<String, String> indirectJarToTarget = new HashMap<>();
+
+  private final Set<String> ignoredTargets;
+  private final Set<String> indirectTargets;
+  private final Set<String> directTargets;
+
+  private CompileOptions ops;
+  public final Reporter delegateReporter;
+  private Set<String> astUsedJars = new HashSet<>();
+
+  public DepsTrackingReporter(Settings settings, CompileOptions ops, Reporter delegate) {
+    super(settings);
+    this.ops = ops;
+    this.delegateReporter = delegate;
+
+    if (ops.directTargets.length == ops.directJars.length) {
+      for (int i = 0; i < ops.directJars.length; i++) {
+        try {
+          jarToTarget.put(ops.directJars[i], ops.directTargets[i]);
+        } catch (ArrayIndexOutOfBoundsException e) {
+          throw new RuntimeException(
+              "mismatched size: " + ops.directJars.length + " vs " + ops.directTargets.length);
+        }
+      }
+    }
+
+    if (ops.indirectTargets.length == ops.indirectJars.length) {
+      for (int i = 0; i < ops.indirectJars.length; i++) {
+        indirectJarToTarget.put(ops.indirectJars[i], ops.indirectTargets[i]);
+      }
+    }
+
+    ignoredTargets = Arrays.stream(ops.unusedDepsIgnoredTargets).collect(Collectors.toSet());
+    directTargets = Arrays.stream(ops.directTargets).collect(Collectors.toSet());
+    indirectTargets = Arrays.stream(ops.indirectTargets).collect(Collectors.toSet());
+  }
+
+  private boolean isDependecyTrackingOn() {
+    return Objects.equals(ops.dependencyTrackingMethod, "ast-plus")
+        && (!"off".equals(ops.strictDepsMode) || !"off".equals(ops.unusedDependencyCheckerMode));
+  }
+
+  @Override
+  public void info0(Position pos, String msg, Severity severity, boolean force) {
+    if (msg.startsWith("DT:")) {
+      if (isDependecyTrackingOn()) {
+        parseOpenedJar(msg);
+      }
+    } else {
+      if (delegateReporter != null) {
+        delegateReporter.info0(pos, msg, severity, force);
+      } else {
+        super.info0(pos, msg, severity, force);
+      }
+    }
+  }
+
+  private void parseOpenedJar(String msg) {
+    usedJars.add(msg.split(":")[1]);
+  }
+
+  public void prepareReport() throws IOException {
+    Map<String, Dep> usedTargets = new HashMap<>();
+
+    for (String jar : usedJars) {
+      String target = jarToTarget.get(jar);
+      boolean isDirect = target != null;
+
+      if (target == null) {
+        target = indirectJarToTarget.get(jar);
+      }
+
+      if (target.startsWith("Unknown")) {
```

Comment:
I ended up with target being null here on scala 2.11

Reply:
This seems to happen from jars from the jdk for me --
external/remote_jdk8_macos_aarch64/zulu-8.jdk/Contents/Home/jre/lib/rt.jar is triggering this, since its not a direct or indirect target

Reply:
(i did go back now and just try confirm unused deps in ast mode works, its too buggy on 2.11 for us to use. But it does run. [this i believe is a known thing that its not great on 2.11 going back many years.... some day we'll get this codebase on 2.12....])

Reply:
I think the null check just needs to be moved up a few lines, looks like you have one one right after this

Reply:
Thanks, @ianoc, this is a good catch! I think I need to ignore jars which are not associated with labels as it is done in the analyzer plugin.

Reply:
I hope https://github.com/bazelbuild/rules_scala/pull/1459/commits/27f96d48c00e5556db148969131853b9b65073a1 fixes the problem, but I haven't tested it yet